### ISSUE TITLE
Crash with @RequiresDetToString and varargs

### ISSUE THREAD

**Issue by jwaataja** [t-rasmud/checker-framework#137]
I get a crash when calling methods annotated with `@RequiresDetToString` that take variable argument lists. Consider the following class,
```java
import org.checkerframework.checker.determinism.qual.*;
import java.util.*;

class C {
    @RequiresDetToString
    public static void f(String... args) {
    }

    public static void g() {
        f("a", "b");
    }
}
```

I get the following error,
```
error: SourceChecker.typeProcess: unexpected Throwable (IndexOutOfBoundsException) while processing C.java; message: Index: 1, Size: 1
  Compilation unit: C.java
  Last visited tree at line 10 column 9:
          f("a", "b");
  Exception: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1; Stack trace: jdk.compiler/com.sun.tools.javac.util.List.get(List.java:490)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethodInvocation(DeterminismVisitor.java:529)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethodInvocation(DeterminismVisitor.java:31)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1650)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:433)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1460)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1032)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  jdk.compiler/com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
  org.checkerframework.framework.source.SourceVisitor.visitMethod(SourceVisitor.java:106)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:750)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethod(DeterminismVisitor.java:421)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethod(DeterminismVisitor.java:31)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
  jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
  org.checkerframework.framework.source.SourceVisitor.visitClass(SourceVisitor.java:94)
  org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:383)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:336)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1020)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1410)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1367)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:965)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:306)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:165)
  jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
  jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
```

From inspecting the code in the stack trace, this seems to be because there's a check that iterates over each parameter using the size of the arguments, but the sizes of parameters and arguments differ with methods that take a variable number of arguments.

**Comment by t-rasmud** [comment]
Fixed in 70d8368e88d4101f82390e48cd5ec4b63b458f2b

**Comment by jwaataja** [comment]
Now, there's an error when the number of arguments is less than the number of parameters, as in this class,
```java
import org.checkerframework.checker.determinism.qual.*;
import java.util.*;

class C {
    @RequiresDetToString
    public static void f(String... args) {
    }

    public static void g() {
        f();
    }
}
```
which has a similar error
```
error: SourceChecker.typeProcess: unexpected Throwable (IndexOutOfBoundsException) while processing C.java; message: Index: 0, Size: 0
  Compilation unit: C.java
  Last visited tree at line 10 column 9:
          f();
  Exception: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0; Stack trace: jdk.compiler/com.sun.tools.javac.util.List.get(List.java:490)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethodInvocation(DeterminismVisitor.java:529)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethodInvocation(DeterminismVisitor.java:32)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1650)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:433)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1460)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1032)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  jdk.compiler/com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
  org.checkerframework.framework.source.SourceVisitor.visitMethod(SourceVisitor.java:106)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:778)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethod(DeterminismVisitor.java:422)
  org.checkerframework.checker.determinism.DeterminismVisitor.visitMethod(DeterminismVisitor.java:32)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:298)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
  jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
  org.checkerframework.framework.source.SourceVisitor.visitClass(SourceVisitor.java:94)
  org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:383)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:336)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:164)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1028)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1410)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1367)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:965)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:306)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:165)
  jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
  jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
```

Looking at the code that causes this, it's [this snippet](https://github.com/t-rasmud/checker-framework/blob/b3957cf1bee3b6fb785d8a052a354be8885c62ba/checker/src/main/java/org/checkerframework/checker/determinism/DeterminismVisitor.java#L528).
```java
        for (int index = 0; index < params.size(); index++) {
            ExpressionTree arg = args.get(index);
            VariableElement param = params.get(index);
```
Before we used `args.size()` rather than `params.size()`. I don't think it would be enough to change the check to `index < params.size() && index < args.size()` because then we would miss checking every argument. The problem is that with a variable length argument list, the last parameter corresponds to 0 or more arguments. So, perhaps the check should be `index < args.size()` but then we change
```java
VariableElement param = params.get(index);
```
to
```java
VariableElement param = params.get(Math.min(index, params.size() - 1))
```
or something to that effect.

**Comment by t-rasmud** [comment]
Fixed in  32d3130
